src/In.ProjectEKA.HipService/Discovery/PatientDiscovery.cs

```diff
@@ -1,39 +1,78 @@
+        private readonly ILinkPatientRepository linkPatientRepository;
+        private readonly IPatientRepository patientRepository;
 
         public PatientDiscovery(
-            IMatchingRepository patientRepository,
-            IDiscoveryRequestRepository discoveryRequestRepository)
+            IMatchingRepository matchingRepository,
+            IDiscoveryRequestRepository discoveryRequestRepository,
+            ILinkPatientRepository linkPatientRepository,
+            IPatientRepository patientRepository)
         {
-            matchingRepository = patientRepository;
+            this.matchingRepository = matchingRepository;
             this.discoveryRequestRepository = discoveryRequestRepository;
+            this.linkPatientRepository = linkPatientRepository;
+            this.patientRepository = patientRepository;
             filter = new Filter();
         }
 
         public async Task<Tuple<DiscoveryRepresentation, ErrorRepresentation>> PatientFor(DiscoveryRequest request)
         {
-            var patients = await matchingRepository.Where(request);
-            var (patient, error) = DiscoveryUseCase.DiscoverPatient(filter.Do(patients, request).AsQueryable());
+            var (linkRequest, exception) = await linkPatientRepository.GetLinkedCareContexts(request.Patient.Id);
+            if (exception != null)
+            {
+                return new Tuple<DiscoveryRepresentation, ErrorRepresentation>(null,
+                    new ErrorRepresentation(new Error(ErrorCode.FailedToGetLinkedCareContexts,
+                        "Failed to get Linked Care Contexts")));
+            }
+
+            if (linkRequest != null)
+            {
+                return patientRepository.PatientWith(linkRequest.PatientReferenceNumber)
+                    .Map(patient => new Tuple<DiscoveryRepresentation, ErrorRepresentation>(
+                        new DiscoveryRepresentation(patient.ToPatientEnquiryRepresentation(
+                            GetUnlinkedCareContexts(linkRequest, patient))),
+                        null)
+                    ).ValueOr(
+                        new Tuple<DiscoveryRepresentation, ErrorRepresentation>(
+                            null,
+                            new ErrorRepresentation(new Error(ErrorCode.NoPatientFound, ErrorMessage.NoPatientFound))));
+            }
 
-            if (patient == null)
+            var patients = await matchingRepository.Where(request);
+            var (patientEnquiryRepresentation, error) = DiscoveryUseCase.DiscoverPatient(
+                filter.Do(patients, request).AsQueryable());
+            if (patientEnquiryRepresentation == null)
             {
                 return new Tuple<DiscoveryRepresentation, ErrorRepresentation>(null, error);
             }
 
             await discoveryRequestRepository.Add(new Model.DiscoveryRequest(request.TransactionId,
                 request.Patient.Id));

```

Comment:
Don't we have to do the same thing for the discovering unlinked care context also, otherwise link will fail no?